### ISSUE TITLE
Modified beam search LODR on an overfit RNLMM

### ISSUE THREAD

**Issue by tattorba87** [k2-fsa/icefall#1282]
Hello, 

I have been attempting to use modified beam search LODR decoding `--decoding-method modified_beam_search_LODR` with an RNLLM trained on text in the target domain which is a small set of voice transcripts I reserved for this (~ 9hours or 65K words in the transcripts). The RNNLM is relatively small due to the size of the data `--rnn-lm-embedding-dim 64 --rnn-lm-hidden-dim 32 --rnn-lm-num-layers 2 `. I use a bigram model trained on the transcripts used to train a zipformer model.  The problem is that I am practically getting no gains in WER out of this, or very small relative improvements (1.5%).

So I thought I would try and debug my pipeline in case I did something wrong by training the RNNLM directly on the testset (5% is left out for validation), such that it intentionally overfits to the test set. I used a larger RNLLM to ensure that it does indeed overfit. At the end of training perplexity is at: tot_ppl ~ 2.0 and valid_ppl ~ 2304. So the perplexity on the testset is where I want it to be. 

The odd thing is when I try to decode once again with lm-scales varying across ( 0.25 0.5 0.75 1.0 ) and I set `--ngram-lm-scale 0` (to simplify matters to begin with) I get terrible results, which seem to get worse as the lm-scale increases!

baseline WER = 29.16 (what I am hoping to improve over with LODR decoding)

Beam | lm_scale | ngram_lm_scale | lm-avg | emb-dim | hid-dim | hid-layers | WER
-- | -- | -- | -- | -- | -- | -- | --
4 | 0.25 | 0.0 | 1 | 400 | 150 | 2 | 36.32
4 | 0.50 | 0.0 | 1 | 400 | 150 | 2 | 54.89
4 | 0.75 | 0.0 | 1 | 400 | 150 | 2 | 61.86
4 | 1.00 | 0.0 | 1 | 400 | 150 | 2 | 66.55

Am I missing something here. This is unexpected right? If the RNNLM is practically memorizing the testset then the WER should drop significantly as the lm-scale increases.

Any ideas to debug the problem are most welcome.  

Thanks :)

**Comment by marcoyang1998** [comment]
In you overfitting experiments with RNNLM, have you checked the error patterns?

**Comment by tattorba87** [comment]
@marcoyang1998 thanks - I did take a quick look at this, what I can say is that there are mostly deletion errors:
```
%WER = 36.32
Errors: 374 insertions, 4740 deletions, 1946 substitutions, over 19440 reference words (12754 correct)
```

**Comment by ezerhouni** [comment]
@tattorba87 Have you tried negative ``lm_scale`` ?

**Comment by tattorba87** [comment]
@ezerhouni What is the thinking here? Why would negative lm-scale help? But I can try it. 
Maybe negative --ngram-lm-scale will help in case the internal LM is interfering in some way.

**Comment by marcoyang1998** [comment]
@tattorba87 BTW, `--ngram-lm-scale` should be negative when doing LODR.
